### ISSUE TITLE
Requirement: Machine-readable maintainers file in collections

### ISSUE THREAD

**Issue by mariolenz** [ansible-community/community-topics#100]
### Summary

For new collections, the maintainers are documented in `collection-meta.yaml`. However, this is a manual process. This is not only error-prone, I also fear that changes to the list of maintainers aren't (always) communicated to us so that information can easily get out of date.

Therefor, I wanted to ask what you think about requiering a machine-readable maintainers file in the collections that are part of the community package. This way:

1. we wouldn't have to add the maintainers manually
2. the collections could update the list of maintainers themselves

It would be nice to have a `maintainers` key in `galaxy.yml`. But this would mean a change to `ansible-core`. Alternatively, we could define our own file of maintainers, `MAINTAINERS.yml` or something.

Thoughts?

### Additional Information

See the comments in ansible-community/ansible-build-data#120

**Comment by briantist** [comment]
I'm fine with this, no particularly strong feelings.

I'm not sure about separate file vs. `galaxy.yml`. I kind of feel like, if we keep shying away from making relevant changes in core, then we're constantly needing to create new files and workaround it. `galaxy.yml` feels like the correct place.

Or maybe we should define a single file like `community.yml` or whatever for things that are community package-related, but don't have any dependency on core?

Really not sure what's best... but implementation details aside I'm +1 on machine-readable maintainers.

**Comment by gundalow** [comment]
+1 to what this is trying to achive

I'm not sure if `galaxy.yml` ends up in the final collection. I think it gets rewritten as `galaxy.json`

I like @briantist's idea of `community.yml` to avoid more files in the root, maybe we could use the existing `meta` directory (currently has `runtime.yml` in), so we'd have `meta/community.yml`

**Comment by mariolenz** [comment]
> I'm not sure about separate file vs. `galaxy.yml`. I kind of feel like, if we keep shying away from making relevant changes in core, then we're constantly needing to create new files and workaround it.

Good point.

> `galaxy.yml` feels like the correct place.

For me, too. I just don't know how easy it is to add changes to the `galaxy.yml` structure in core. But yes, it feels like the best place for documenting the maintainers.

> Or maybe we should define a single file like `community.yml` or whatever for things that are community package-related, but don't have any dependency on core?

Good idea! A `MAINTAINERS.yml` sounds quite specific. `community.yml` would be more generic. But since the community package is just ansible-core plus collections, it would be better if we could get all we need from the `galaxy.yml` without the need for additional files.

**Comment by briantist** [comment]
If we did do a separate file, I also agree I'd like to see it in `meta/` and not the root.

**Comment by felixfontein** [comment]
> For me, too. I just don't know how easy it is to add changes to the `galaxy.yml` structure in core. But yes, it feels like the best place for documenting the maintainers.

I think that'll be pretty hard, since both galaxy.yml *and* MANIFEST.json have to be adjusted, and this probably also affects Galaxy and Galaxy NG, and maybe even other RH products.

Having `meta/community.yml` sounds good. One question about it would be who should lint it though. We'll definitely need a linter to avoid chaos :) But having this in ansible-core's ansible-test isn't a good idea (it would be hard to add new things, if you want to lint strictly), and a lint subcommand of antsibull-docs isn't a great place either. Once ansible-lint can do such things it is probably a good place to do that, but I don't know what's its current state.

**Comment by felixfontein** [comment]
> > a lint subcommand of antsibull-docs isn't a great place either. Once ansible-lint can do such things it is probably a good place to do that, but I don't know what's its current state.
> 
> What about a GitHub workflow, at least to start with? Similar to [ansible-community/ansible-build-data#122](https://github.com/ansible-community/ansible-build-data/pull/122)?

I think we want something that the collections themselves can easily run to determine whether they conform, instead of us finding that out only after they made a release with the bad info.

**Comment by schurzi** [comment]
+1 for the idea.

Regarding file location I would say, that `galaxy.yml` is the intuitive place from my point of view. Especially since there are already elements for `author`, `repository` and `issues`. These things are semantically connected and placing `maintainers` in a separate file feels confusing to me.

> What about a GitHub workflow, at least to start with? Similar to https://github.com/ansible-community/ansible-build-data/pull/122?

I like that idea too. GitHub has a functionality to include workflows from other repositories (https://docs.github.com/en/actions/using-workflows/reusing-workflows). There are several workflows that come to my mind, which could be supplied by the Ansible community (testing, linting, packaging, ...).

**Comment by jborean93** [comment]
I'm not sure using `galaxy.yml` is a good idea for this. Typically it contains the information used to build a collection. We somewhat have a precedence to use `meta/...` to store metadata about a collection and keep it away from `galaxy.yml` which is for builting a collection.

**Comment by felixfontein** [comment]
If the information is only in galaxy.yml and not transferred into MANIFEST.json, we'd have to fetch this information from the git repository (before the collection is built). And not every collection necessarily has a git repo (we also allow other SCMs), and not every collcetion repo has a galaxy.yml (they could be templated or somehow else generated during the build process by some scripts).

> What about a GitHub workflow, at least to start with? Similar to [ansible-community/ansible-build-data#122](https://github.com/ansible-community/ansible-build-data/pull/122)?

One big problem with that is that it only works with GitHub, and you cannot run it locally or anywhere else. Linting should not be tied that much to a specific hosting platform.

The upcoming (?) 'profiles' feature of ansible-lint might help here though.

**Comment by mariolenz** [comment]
> If the information is only in galaxy.yml and not transferred into MANIFEST.json, we'd have to fetch this information from the git repository (before the collection is built). And not every collection necessarily has a git repo (we also allow other SCMs)

We did allow other SCMs, but we don't do now: https://github.com/ansible-collections/overview/blob/5c03b3af036c4ea3ce7db22040f5fc99b2e8a77c/collection_requirements.rst#repository-management

**Comment by Andersson007** [comment]
How about requiring [CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#codeowners-file-location) file? Moreover, in this case, GitHub will automatically assigned issues/PRs. It's also supported by gitlab as far as i can see.

**Comment by felixfontein** [comment]
Hmm, that would collide with .github/BOTMETA.yml, which is used by the bot. I guess it would be possible to mirate to using that, but first the bot would need to support it.

**Comment by Andersson007** [comment]
> Hmm, that would collide with .github/BOTMETA.yml, which is used by the bot. I guess it would be possible to mirate to using that, but first the bot would need to support it.

@felixfontein how about BOTMETA.yml OR CODEOWNERS? i.e. one of.

**Comment by mariolenz** [comment]
@nitzmahone

Seeing [your comment on How to mark private plugins in a collection](https://github.com/ansible-community/community-topics/issues/154#issuecomment-1464590465), I wonder what you think about using core-owned structures to document maintainers.

I still feel that `galaxy.yml` would be a good place. Maybe you agree, maybe you think that maintainers should be documented in another core-owned structure, maybe you think this should be done somewhere else. It would be really great to get your opinion on this.

cc @bcoca

**Comment by felixfontein** [comment]
I think something like `meta/ext.yml` / `meta/extra.yml` / $BIKESHEDDEDNAME would be better for such information; if some namespaced solution is used as described in https://github.com/ansible-community/community-topics/issues/154#issuecomment-1465108005, we could have a rigid specification how that information is stored in the community part, including having a linter for the community-namespaced things in that file.

**Comment by mariolenz** [comment]
@felixfontein Why? `galaxy.yml` already has an `authors` key (which I don't think is really well defined).

It has keys for `repository`, `documentation`, `homepage` and `issues`. So why not for `maintainers`?

Anyway, I think we have to work closer together with core on questions like this. I sometimes think we, the community, don't talk enough with the core team (and, maybe, they not enough with us). So I'd like to know what they think about this. That's why I've mentioned @nitzmahone and @bcoca (I think they're both core).

**Comment by felixfontein** [comment]
> @felixfontein Why? `galaxy.yml` already has an `authors` key (which I don't think is really well defined).

Yes, `authors` is a list with free-form strings.

> It has keys for `repository`, `documentation`, `homepage` and `issues`. So why not for `maintainers`?

That can be added, *but* for that be useful it also needs to be added to MANIFEST.json, so the data is kept after building the collection, which in turn requires coordination with a lot of entities. And if I learned something from the semantic markup process, this is something that can take forever. So if you want to have something soon, only going the galaxy.yml/MAINFEST.json route is not a good idea.

Also if you want something like `maintainers` in galaxy.yml/MANIFEST.json, it will likely be another (list of) free-form string(s), since it doesn't make sense to add validation for a very strict use-case parts of the community have to core. So if you want something that conforms to a strict syntax, galaxy.yml/MANIFEST.json sounds like the wrong approach to me.

**Comment by bcoca** [comment]
`authors` is the intended field for maintainers, the `author` field spec is already well defined on the module/plugin side, though I don't believe we validate it the same on the galaxy.yml the same way (we should).

**Comment by briantist** [comment]
> `authors` is the intended field for maintainers, the `author` field spec is already well defined on the module/plugin side, though I don't believe we validate it the same on the galaxy.yml the same way (we should).

In my mind the difference is that an author is always an author.

Maintainers change over time.

**Comment by mariolenz** [comment]
Thanks for clarifying this @bcoca!

This sounds like it wouldn't be a good idea to create another place to document maintainers. Having the same information in different, unrelated places looks like an accident waiting to happen.

Although I kind of agree with @briantist, the decision to use the `authors` keyword to document maintainers looks a bit unfortunate to me.

> the `author` field spec is already well defined on the module/plugin side, though I don't believe we validate it the same on the galaxy.yml the same way (we should).

Does "we should" mean there are plans to implement this? Or, at least, that the core team is open to do it?

> So if you want to have something soon, only going the galaxy.yml/MAINFEST.json route is not a good idea.

Well, I've opened this issue 10 months ago so I'd say the point to get this soon has long been gone :laughing: 

Anyway, since `authors` in `galaxy.yml` is the intended field for maintainers (see above) I don't think we should duplicate the information somewhere else.

**Comment by felixfontein** [comment]
`authors` in galaxy.yml might have been intended that way, but it was never used that way and never documented that way, so I would say that it is **not** for maintainers, but for authors. Same as the `authors` field in modules/plugins/roles.
